packages/expo-cli/src/commands/run/android/runAndroid.ts

```diff
@@ -26,28 +36,88 @@ async function resolveAndroidProjectPathAsync(projectRoot: string): Promise<stri
   }
 }
 
-export async function runAndroidActionAsync(projectRoot: string, options: Options) {
-  if (typeof options.buildVariant !== 'string') {
-    throw new CommandError('--build-variant must be a string');
+async function resolveOptionsAsync(
+  projectRoot: string,
+  options: Options
+): Promise<AndroidRunOptions> {
+  if (typeof options.variant !== 'string') {
+    throw new CommandError('--variant must be a string');
   }
+  const device = await resolveDeviceAsync(options.device);
+  if (!device) {
+    throw new CommandError('Cannot resolve an Android device');
+  }
+
+  const filePath = await AndroidConfig.Paths.getAndroidManifestAsync(projectRoot);
+  const androidManifest = await AndroidConfig.Manifest.readAndroidManifestAsync(filePath);
 
-  const bootedDevice = await resolveDeviceAsync(options.device);
-  if (!bootedDevice) {
-    return;
+  // Assert MainActivity defined.
+  await AndroidConfig.Manifest.getMainActivityOrThrow(androidManifest);
+  const mainActivity = 'MainActivity';
+  const packageName = androidManifest.manifest.$.package;
+
+  if (!packageName) {
+    throw new CommandError(`Could not find package name in AndroidManifest.xml at "${filePath}"`);
   }
+  const variant = options.variant.toLowerCase();
+  const apkDirectory = Android.getAPKDirectory(projectRoot);
+  const apkVariantDirectory = path.join(apkDirectory, variant);
+
+  return {
+    ...options,
+    device,
+    mainActivity,
+    packageName,
+    apkVariantDirectory,
+    variantFolder: variant,
+    appName: 'app',
+  };
+}
+
+export async function runAndroidActionAsync(projectRoot: string, options: Options) {
+  const props = await resolveOptionsAsync(projectRoot, options);
 
   Log.log('Building app...');
 
   const androidProjectPath = await resolveAndroidProjectPathAsync(projectRoot);
 
-  await Gradle.spawnGradleTaskAsync({ androidProjectPath, buildVariant: options.buildVariant });
+  await spawnGradleAsync({ androidProjectPath, variant: options.variant });
+
+  const apkFile = await getInstallApkNameAsync(props.device, props);
+  Log.debug(`Installing: ${apkFile}`);
+  const binaryPath = path.join(props.apkVariantDirectory, apkFile);
+  await Android.installOnDeviceAsync(props.device, { binaryPath });
```

Comment:
If we're going to use `adb install` to get the APK installed on the device, maybe switch to a different Gradle task that doesn't install the APK? (Now we run `gradlew installDebug`/`gradlew installRelease` (https://developer.android.com/studio/build/building-cmdline#build_apk) to build and install the app, followed by `adb install`, which also installs the app.)